pkg/vpcmodel/groupingSelfLoop.go

```diff
@@ -0,0 +1,335 @@
+	//    2.1 go over the new data structure, merge groups to be merged and add to New
+	//    2.2 go over old map[string][]*GroupedConnLine and for each element whose key not in toMergeKeys then just add it as is
+	for _, toBeMergedKeys := range toMergeList {
+		newKey, newGroupedConnLines := mergeGivenList(oldGroupingSrcOrDst, srcGrouping, toBeMergedKeys)
+		mergedGroupedConnLine[newKey] = newGroupedConnLines
+	}
+	for oldKey, oldLines := range oldGroupingSrcOrDst {
+		// not merged with other groups, add as is
+		if _, ok := toMergeExistingIndexes[oldKey]; !ok {
+			mergedGroupedConnLine[oldKey] = oldLines
+		}
+	}
+	return mergedGroupedConnLine
+}
+
+// given a list of keys to be merged from of oldGroupingSrcOrDst, computes unique list of endpoints
+// of either sources or destination as by srcGrouping
+// returns the unique list of endpoints, their names and the connection
+func listOfUniqueEndpoints(oldGroupingSrcOrDst map[string][]*GroupedConnLine, srcGrouping bool,
+	toMergeKeys []string) (listOfEndpoints groupedEndpointsElems, setOfNames map[string]struct{}, conn string) {
+	setOfNames = make(map[string]struct{})
+	listOfEndpoints = make(groupedEndpointsElems, 0)
+	for _, oldKeyToMerge := range toMergeKeys {
+		for _, line := range oldGroupingSrcOrDst[oldKeyToMerge] {
+			endPointInKey := line.getSrcOrDst(!srcGrouping)
+			if conn == "" {
+				conn = line.Conn // connection is the same for all lines to be merged
+			}
+			if _, isSliceEndpoints := endPointInKey.(*groupedEndpointsElems); isSliceEndpoints {
+				for _, endpoint := range *endPointInKey.(*groupedEndpointsElems) {
+					if _, ok := setOfNames[endpoint.Name()]; !ok { // was endpoint added already?
+						listOfEndpoints = append(listOfEndpoints, endpoint)
+						setOfNames[endpoint.Name()] = struct{}{}
+					}
+				}
+			} else { // endpoint is Node or NodeSet
+				if _, ok := setOfNames[endPointInKey.Name()]; !ok { // was endpoint added already?
+					listOfEndpoints = append(listOfEndpoints, endPointInKey)
+					setOfNames[endPointInKey.Name()] = struct{}{}
+				}
+			}
+		}
+	}
+	return
+}
+
+func mergeGivenList(oldGroupingSrcOrDst map[string][]*GroupedConnLine, srcGrouping bool,
+	toMergeKeys []string) (newKey string, newGroupedConnLine []*GroupedConnLine) {
+	epsInNewKey, namesInNewKey, _ := listOfUniqueEndpoints(oldGroupingSrcOrDst, srcGrouping, toMergeKeys)
+	epsInNewLines, _, conn := listOfUniqueEndpoints(oldGroupingSrcOrDst, !srcGrouping, toMergeKeys)
+	for _, epInLineValue := range epsInNewLines {
+		if srcGrouping {
+			newGroupedConnLine = append(newGroupedConnLine, &GroupedConnLine{epInLineValue, &epsInNewKey, conn})
+		} else {
+			newGroupedConnLine = append(newGroupedConnLine, &GroupedConnLine{&epsInNewKey, epInLineValue, conn})
+		}
+	}
+	srcsOrDstsInNewKeySlice := make([]string, 0)
+	for item := range namesInNewKey {
+		srcsOrDstsInNewKeySlice = append(srcsOrDstsInNewKeySlice, item)
+	}
+	newKey = strings.Join(srcsOrDstsInNewKeySlice, commaSeparator) + conn
```

Comment:
The format of the key here is a bit different from the original format (see line 233 in `grouping.go`). 
Consider adding a function to handle the key's format uniquely in all places.

Reply:
Seems line 233 in `grouping.go` has shifted. To be on the safe side, please refer to the relevant code (not line)

Reply:
the line I referred to: `key := dstOrSrc.Name() + ";" + line.Conn`

Reply:
I'm not sure I am following. Perhaps we should take it offline

Reply:
Groups are merged only after it was already verified they have the same connection. Hence it makes no sense to add the connection to the key. This is different than the lines you quota from `grouping.go`

Reply:
I meant that you don't add `;` between the different parts in the key. In you version it is `strings.Join(srcsOrDstsInNewKeySlice, commaSeparator) + conn` , and in the line I was referring to, the `;` is added before the `conn` part.
I think the format should be the same.

Reply:
Ahh, now I understand.